packages/dashboard/src/components/iot-dashboard/iot-dashboard-internal.tsx

```diff
@@ -0,0 +1,443 @@
+      .widgets.filter(({ id }) => !existingWidgetIds.includes(id))
+      .map(({ id }) => id);
+    this.selectWidgets({
+      widgetIds: newlyCreatedWidgetIds,
+    });
+  }
+
+  /**
+   *
+   * Gesture Start
+   *
+   */
+
+  onGestureStart(event: MouseEvent) {
+    const { x, y } = getDashboardPosition(event);
+    const isMoveGesture = !event.shiftKey && this.isPositionOnWidget({ x, y });
+
+    if (isMoveGesture) {
+      this.onMoveStart({ x, y });
+    } else {
+      this.onSelectionStart(event);
+    }
+    // NOTE: Resize is initiated within the `<iot-selection-box />`
+  }
+
+  onSelectionStart(event: MouseEvent) {
+    this.activeGesture = 'selection';
+
+    const { x, y } = getDashboardPosition(event);
+    this.start = { x, y };
+    this.end = { x, y };
+
+    const isUnionSelection = event.shiftKey;
+    const intersectedWidgetIds = getSelectedWidgetIds({
+      selectedRect: this.selectedRect(),
+      dashboardConfiguration: this.getDashboardConfiguration(),
+      cellSize: this.cellSize,
+    });
+
+    const newlySelectedWidgetIds = intersectedWidgetIds.filter((id) => !this.selectedWidgetIds.includes(id));
+    this.selectWidgets({
+      widgetIds: isUnionSelection ? [...this.selectedWidgetIds, ...newlySelectedWidgetIds] : intersectedWidgetIds,
+    });
+  }
+
+  onResizeStart: OnResize = ({ anchor, currentPosition }) => {
+    this.activeGesture = 'resize';
+
+    this.activeResizeAnchor = anchor;
+    this.resizeStartPosition = currentPosition;
+    this.startResize = currentPosition;
+  };
+
+  onMoveStart({ x, y }: Position) {
+    this.activeGesture = 'move';
+    this.startMove = { x, y };
+    const intersectedWidgetIds = getSelectedWidgetIds({
+      selectedRect: { x, y, width: 1, height: 1 },
+      dashboardConfiguration: this.getDashboardConfiguration(),
+      cellSize: this.cellSize,
+    });
+
+    const selectingAlreadySelectedWidget = intersectedWidgetIds.some((widgetId) =>
+      this.selectedWidgetIds.includes(widgetId)
+    );
+
+    if (!selectingAlreadySelectedWidget) {
+      this.setSelectedWidgets();
+    }
+
+    this.previousPosition = { x, y };
+    if (this.selectedWidgetIds.length === 0) {
+      this.moveStartWidgetIds = intersectedWidgetIds;
+      this.selectWidgets({
+        widgetIds: intersectedWidgetIds,
+      });
+    }
+  }
+
+  /**
+   *
+   * On gesture update
+   *
+   */
+
+  onGestureUpdate(event: MouseEvent) {
+    if (this.activeGesture === 'move') {
+      this.onMove(getDashboardPosition(event));
+    } else if (this.activeGesture === 'resize') {
+      this.onResize(event);
+    } else if (this.activeGesture === 'selection') {
+      this.onSelection(event);
+    }
+  }
+
+  onMove({ x, y }: Position) {
+    if (this.previousPosition) {
+      this.move({
+        position: { x, y },
+        prevPosition: this.previousPosition,
+        widgetIds: this.moveStartWidgetIds,
+        isActionComplete: false,
+      });
+
+      this.previousPosition = { x, y };
```

Comment:
could `previousPosition` be moved into the reducer as well and make it so the dashboard doesn't need to worry about tracking it?